File: sdks/python/src/backlex/realtime.py
"""SSE realtime transport.

The backlex realtime plane is Server-Sent Events (not WebSockets), so this is a
minimal SSE reader running on a daemon thread. ``subscribe`` returns an
``unsubscribe`` callable — the same ``() -> None`` contract as the TS SDK. The
reader auto-reconnects on a dropped stream (3s back-off, matching the server's
reconnect hint) and replays via ``Last-Event-ID`` when the server supplies ids.
"""

from __future__ import annotations

import json
import threading
from typing import Any, Callable, Dict, Optional

import httpx

from .errors import BacklexError
from .types import ItemEvent

OnEvent = Callable[[ItemEvent], None]
OnError = Callable[[Any], None]
Unsubscribe = Callable[[], None]

_RECONNECT_SECONDS = 3.0


def subscribe(
    http: httpx.Client,
    url: str,
    auth_header: Callable[[], Dict[str, str]],
    on_event: OnEvent,
    on_error: Optional[OnError] = None,
) -> Unsubscribe:
    stop = threading.Event()

    def run() -> None:
        last_id: Optional[str] = None
        while not stop.is_set():
            headers = {"accept": "text/event-stream", **auth_header()}
            if last_id is not None:
                headers["last-event-id"] = last_id
            try:
                with http.stream("GET", url, headers=headers, timeout=None) as resp:
                    if resp.status_code != 200:
                        raise BacklexError(resp.status_code, None)
                    data_lines: list[str] = []
                    for line in resp.iter_lines():
                        if stop.is_set():
                            return
                        if line == "":
                            # Blank line dispatches the buffered event.
                            if data_lines:
                                payload = "\n".join(data_lines)
                                data_lines = []
                                try:
                                    on_event(json.loads(payload))
                                except Exception as exc:  # noqa: BLE001
                                    if on_error:
                                        on_error(exc)
                            continue
                        if line.startswith(":"):
                            # Comment / heartbeat frame.
                            continue
                        if line.startswith("id:"):
                            last_id = line[3:].strip()
                            continue
                        if line.startswith("data:"):
                            data_lines.append(line[5:].lstrip())
            except Exception as exc:  # noqa: BLE001
                if stop.is_set():
                    return
                if on_error:
                    on_error(exc)
            # Reconnect after a short back-off unless we've been told to stop.
            stop.wait(_RECONNECT_SECONDS)

    thread = threading.Thread(target=run, name=f"backlex-sse:{url}", daemon=True)
    thread.start()

    def unsubscribe() -> None:
        stop.set()

    return unsubscribe

File: sdks/ruby/lib/backlex/realtime.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"

module Backlex
  # Handle for an active realtime subscription. #close unsubscribes — the same
  # contract as the TS SDK's returned unsubscribe function. The reader runs on a
  # background thread and auto-reconnects on a dropped stream (3s back-off),
  # replaying via Last-Event-ID.
  class Subscription
    def initialize(client, url, on_event, on_error)
      @stopped = false
      @thread = Thread.new { run(client, url, on_event, on_error) }
    end

    def close
      @stopped = true
      @thread&.kill
    end

    private

    def run(client, url, on_event, on_error)
      last_id = nil
      until @stopped
        begin
          uri = URI(url)
          Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") do |http|
            req = Net::HTTP::Get.new(uri)
            req["Accept"] = "text/event-stream"
            client.auth_header(req)
            req["Last-Event-ID"] = last_id if last_id
            http.request(req) do |res|
              if res.code.to_i != 200
                on_error&.call(Backlex::Error.new(res.code.to_i, "UNKNOWN", "HTTP #{res.code}"))
                next
              end
              last_id = read_stream(res, on_event, on_error, last_id)
            end
          end
        rescue StandardError => e
          on_error&.call(e) unless @stopped
        end
        break if @stopped

        sleep 3
      end
    end

    # Parse the SSE byte stream, buffering across chunk boundaries.
    def read_stream(res, on_event, on_error, last_id)
      buffer = +""
      data = []
      res.read_body do |chunk|
        break if @stopped

        buffer << chunk
        while (idx = buffer.index("\n"))
          line = buffer.slice!(0, idx + 1).chomp
          if line.empty?
            unless data.empty?
              payload = data.join("\n")
              data = []
              begin
                on_event.call(JSON.parse(payload))
              rescue StandardError => e
                on_error&.call(e)
              end
            end
          elsif line.start_with?(":")
            # comment / heartbeat
          elsif line.start_with?("id:")
            last_id = line[3..].strip
          elsif line.start_with?("data:")
            d = line[5..]
            d = d[1..] if d.start_with?(" ")
            data << d
          end
        end
      end
      last_id
    end
  end
end
